### PR TITLE
Attempt to make support h-matchers multiple python versions

### PR DESCRIPTION
**Pull request by jon-betts**
For: https://github.com/hypothesis/h-matchers/issues/36

Add support for Python 3.7, 3.8 and 3.9.

The idea here is to:

 * Get something which passes locally and in CI
 * Demonstrates how we might do it
 * Attempts to make it as templated as possible
 * Provides a platform to ship off into `hdev` or something similar

### CODE REVIEW

File: tox.ini [hunk 1]
```diff
@@ -1,5 +1,5 @@
 [tox]
-envlist = py36-tests
+envlist = py36-tests,py37-tests,py38-tests,py39-tests
```
Comment (seanh) [inline tox.ini:2]: I think this should work:
```suggestion
envlist = {py36,py37,py38,py39}-tests
```

Comment (jon-betts) [reply tox.ini]: Yeah it does. This is how we are generating the versions from the script, although this is happening in the Makefile now

File: .github/workflows/pythonpackage.yml [hunk 2]
```diff
@@ -2,22 +2,40 @@ name: Python package
 on: [push]
 
 jobs:
-  test:
-    name: Test
+  metadata:
```
Comment (jon-betts) [inline .github/workflows/pythonpackage.yml:5]: This metadata phase takes about 6 seconds

File: bin/python_codes.py [hunk 3]
```diff
@@ -0,0 +1,142 @@
+#!/usr/bin/env python
```
Comment (jon-betts) [inline bin/python_codes.py:1]: This will presumably end up squirrelled away in `hdev` in coming tickets

File: setup.cfg [hunk 4]
```diff
@@ -16,6 +16,9 @@ url = https://web.hypothes.is/
 classifiers =
     Development Status :: 2 - Pre-Alpha
     Programming Language :: Python :: 3.6
+    Programming Language :: Python :: 3.7
+    Programming Language :: Python :: 3.8
+    Programming Language :: Python :: 3.9
```
Comment (jon-betts) [inline setup.cfg:21]: I can't see a super easy way to template this in, but it's not all that onerous  to keep these up to date. The script will generate them for convenience.

File: .github/workflows/pythonpackage.yml [hunk 5]
```diff
@@ -37,6 +55,31 @@ jobs:
...
+
+    - name: Setup python
+      uses: actions/setup-python@v2
+      with:
+        python-version: ${{ matrix.python-version }}
+
+    - name: Install dependencies
+      run: |
+        pip install .[tests] cookiecutter
+
```
Comment (jon-betts) [inline .github/workflows/pythonpackage.yml:82]: Some of this is robbed from @marcospri 's PR here: https://github.com/hypothesis/hdev/pull/4

The examples of how to set outputs etc. was cribbed from here: https://github.com/frezbo/dockerfiles/blob/6701b8cffc3b3628ad72b541d2c36ac87e1e9b8f/.github/workflows/ci.yml

File: .python-version [hunk 6]
```diff
@@ -1 +1,4 @@
-3.6.9
+3.9.2 
+3.8.8
+3.7.10
+3.6.13
```
Comment (jon-betts) [inline .python-version:4]: This file now has a lot more meaning than before:

 * The first entry is taken as the version to run linting against
 * You can add `# future` to include a version in `make testall` but not CI
 * You can add `# floating` to have the version vagued up to `3.9.x` in CI
 * You can add both at once, so for example we could have `3.10.0  # future floating`
 * This doesn't currently make much sense as the floating only happens in CI, and currently we don't have the capability to run it but ignore results, but support is there in the version generator

File: bin/install-python.sh [hunk 7]
```diff
@@ -34,6 +34,9 @@ fi
 # Loop over every $python_version in the .python-version file.
 while IFS= read -r python_version
 do
+    # Strip comments from the end of versions
+    python_version=`echo $python_version | sed -e 's/\s*#[^!].*$//'`
+
     # Install this version of Python in pyenv if it's not installed already.
     pyenv install --skip-existing "$python_version"
```
Comment (jon-betts) [inline bin/install-python.sh:41]: We need to include a version of this file in `hdev` ideally so as not to have to copy paste this change into every repo, although I'm not sure how yet.

File: .github/workflows/pythonpackage.yml [hunk 8]
```diff
@@ -2,22 +2,40 @@ name: Python package
...
 
     steps:
     - name: Checkout code
       uses: actions/checkout@v1
 
-    - name: Show git ref
+    - name: Setup python
+      uses: actions/setup-python@v2
+      with:
+        python-version: 3.9.2
```
Comment (seanh) [inline .github/workflows/pythonpackage.yml:18]: So I guess this Python version is used to _run_ `bin/python_codes.py`?

File: .github/workflows/pythonpackage.yml [hunk 9]
```diff
@@ -2,22 +2,40 @@ name: Python package
...
+
+    - name: List Python versions
+      id: list-python-versions
       run: |
-        echo ${{ github.ref }}
+        VERSIONS_JSON=`./bin/python_codes.py --style gha`
+        HEADLINE_VERSION=`./bin/python_codes.py --style headline`
+        echo "::set-output name=python_versions::${VERSIONS_JSON}"
+        echo "::set-output name=headline_version::${HEADLINE_VERSION}"
```
Comment (seanh) [inline .github/workflows/pythonpackage.yml:26]: Whereas the `headline_version` from `bin/python_codes.py --style headline` is used to actually run the `lint`, `package` and `publish` jobs. And the `python_versions` from `bin/python_codes.py --style gha` are used to run the `test` job

File: .github/workflows/pythonpackage.yml [hunk 10]
```diff
@@ -2,22 +2,40 @@ name: Python package
 on: [push]
 
 jobs:
-  test:
-    name: Test
+  metadata:
     runs-on: ubuntu-latest
+    outputs:
+      python_versions: ${{ steps.list-python-versions.outputs.python_versions }}
+      headline_version: ${{ steps.list-python-versions.outputs.headline_version }}
```
Comment (seanh) [inline .github/workflows/pythonpackage.yml:9]: Might call this `headline_python_version` just to be clear

File: Makefile [hunk 11]
```diff
@@ -42,6 +42,10 @@ initialrelease: python
 test: python
 	@tox -q
 
+.PHONY: testall
+testall: python
+	@tox -qe \{`./bin/python_codes.py`\}-tests
```
Comment (seanh) [inline Makefile:47]: We should add `make sure` to the packages as well and have it run `checkformatting lint testall`

File: .python-version (hunk 6 above)
Comment (seanh) [reply .python-version:4]: I wonder whether `# floating` is just the default behaviour that we want all the time

**Review by seanh — CHANGES_REQUESTED** [review]
I'm not sure about the future versions stuff. I think we might want `make sure` and `make testall` to do the same thing as is done on CI. Otherwise they could be failing locally but passing on CI which might just be confusing.

I think we might just want to add a new Python version to `.python-versions`, fix any bugs, and have it be run both locally and on CI. I'm not sure we'll want `make` commands to be running some Python versions locally that don't get run on CI.

Note that you can always do `tox -e py39-tests` manually to run the tests with some future version of Python

File: .python-version (hunk 6 above)
Comment (jon-betts) [reply .python-version:4]: Depends how you feel about whether what you run locally should reflect what happens in CI. The general thrust of the requirements files has been to ensure we run the same thing in both.

Let me know which way to go.

File: .github/workflows/pythonpackage.yml (hunk 8 above)
Comment (jon-betts) [reply .github/workflows/pythonpackage.yml:18]: Yes, we need something to bootstrap it

File: .python-version (hunk 6 above)
Comment (jon-betts) [reply .python-version:4]: On the other hand I'd suggest that if we want it on every one, put it on every one. This model supports any way of working, we don't need to limit it to one.

File: .github/workflows/pythonpackage.yml (hunk 9 above)
Comment (jon-betts) [reply .github/workflows/pythonpackage.yml]: I've removed the variables.

File: .github/workflows/pythonpackage.yml [hunk 12]
```diff
@@ -2,22 +2,40 @@ name: Python package
...
+      uses: actions/setup-python@v2
+      with:
+        python-version: 3.9.2
+
+    - name: List Python versions
+      id: list-python-versions
       run: |
-        echo ${{ github.ref }}
+        VERSIONS_JSON=`./bin/python_codes.py --style gha`
+        HEADLINE_VERSION=`./bin/python_codes.py --style headline`
```
[the remark this answers is not in the document]
Comment (jon-betts) [reply .github/workflows/pythonpackage.yml]: I've done this. It's a bit odd as it usually returns multiple versions.

File: .github/workflows/pythonpackage.yml (hunk 10 above)
Comment (jon-betts) [reply .github/workflows/pythonpackage.yml]: Done.

**Comment by jon-betts** [comment]
As discussed in https://github.com/hypothesis/product-backlog/issues/1199:

> * Optionally target an "aspirational" version of Python we cannot reach yet to:
>   * Give us warning of any additions which break compatibility with that version
>   * ... or give us notice when we become compatible with that version
> * Testing against all versions should be:
>   ...
>   * Required to pass in all but the "aspirational" target if the "asprirational" target version has never passed
>   * Required to pass in all cases if the "aspirational" target has ever passed

The future requirement is a basic way of achieving this. We add future, and then we can be notified locally but CI will continue to pass. If we notice that local tests are passing for the version we can remove the "future" tag and it will be tested in CI achieving the last part.

While we can test adhoc versions by editing the version file temporarily and running a custom tox command but it:

 * Doesn't allow us to record a decision to formally attempt to move to supporting a version (you have to remember it yourself)
 * Is unlikely to ever become part of common practice
 
An improvement to this process would be to have the tests run on CI, but only "warn" on failure, if such a thing is possible.

File: Makefile (hunk 11 above)
Comment (jon-betts) [reply Makefile]: I've added a `make sure` command locally. We can move it to the cookie cutter when other behaviors move out.

**Comment by jon-betts** [comment]
The `# floating` behavior was added to support the commit that @marcospri made for the `hdev` project. I think the way I'd suggest approaching this is:

* If the system can achieve the behavior we want, it's good
* If it also supports other behaviors too that's better
* Because it allows us to change our mind and experiment without changing the system, only the config

So many of the suggestions in the PR above actually are already in place / could be easily. e.g.:

 * "Run the same versions locally and in CI" - `h_matchers` has no `'# future` tagged versions, so this is in place
 * "Have all the items be floating"  - Add `# floating` to all versions

**Review by seanh — APPROVED** [review]

**Pull request merged**